05-part_two.Rmd

```diff
@@ -30,15 +34,72 @@ get_max(c("a", "b", "c"))
 ```
 
-Beyond in-script tests, an analyst may decide to take a more formal approach to testing if the project warrants it, this is commonly referred to as 'unit testing'. These require both writing tests in separate files and using specialised packages that run all tests at once and return success/failure feedback. The `testthat` package is a popular tool to write and run tests in R (more details [here](https://github.com/r-lib/testthat)). Although most of its documentation and functionalities focus on testing in the context of package writing, `testthat` can also be used with regular R projects.  There's a brief overview of unit testing below, but for a more detailed introduction see [here](https://katherinemwood.github.io/post/testthat/) or [here](https://www.r-bloggers.com/automated-testing-with-testthat-in-practice/) for a more in-depth overview of automated testing, including test driven development. 
 
-In `testthat`, tests are called **expectations**. These spell out the expected behaviour of a specific piece of code - typically a function. `testthat` provides many functionalities to define expectactions - see details [here](http://r-pkgs.had.co.nz/tests.html). Expectations falling within a same testing **context** are stored in a same file and all testing files are saved within a dedicated testing directory. Below is an example of a file containing tests checking data integrity. 
+## Formalised testing
+Beyond in-script tests, an analyst may decide to take a more formal approach to testing if the project warrants it. Software development has dedicated testing professionals (called "testers") and there is plenty of literature about software testing on the internet. Much of this is transferable to code based analysis, but not all of it. There are times when it makes sense to make a distinction between a model (which is generally operated by a technical specialist) and an IT product (e.g. a website, which is used by a broader user base). Shiny apps sit somewhere between a model and an IT product so we'll include some specific comments about these.
+
+There are often four types of software testing covered:
+
+* Unit testing (tests the different components or "units" of code using tools such as `testthat` in R).
+* Integration testing (tests the correct flow of data through the components).
+* System testing (tests the overall outputs make sense).
+* Acceptance testing (tests whether or not the work meets its objectives).
+
+Software testing can make a distinction between functional and non-functional testing too. In the context of an analytical model, a functional test would be one which tests the model produces what its intended to whereas a non-functional test would cover things such as its run-time and interface (i.e. things which affect the experience of using it, but not the numbers that come out).
+
+For more information, see [here](https://www.guru99.com/levels-of-testing.html) or use your favourite search engine to explore the topic.
```

Comment:
Can we bring this sentence up into the previous paragraph and amend is so that it's along the lines of "You can find more information here" so that it's consistent with the writing style of the whole document

Reply:
I've brought this into the paragraph but I've not added "you" in given I was on a spree of removing references to you. Having scanned other sections they don't seem to particularly use "you" when referring to other sources. Please let me know if you think I've missed something.